sa/sa.go

```diff
@@ -915,3 +915,188 @@ func (ssa *SQLStorageAuthority) Health(ctx context.Context) error {
+		}
+
+		// Determine which shard index we want to lease: by default the oldest, but
+		// an arbitrary missing one if any are missing.
+		shardIdx := oldest.Idx
+		needToInsert := false
+		for i := req.MaxShardIdx; i >= req.MinShardIdx; i-- {
+			_, ok := shardMap[int(i)]
+			if !ok {
+				shardIdx = int(i)
+				needToInsert = true
+				break
+			}
+		}
+
+		if needToInsert {
+			_, err = txWithCtx.Exec(
+				`INSERT INTO crlShards (issuerID, idx, leasedUntil)
+				VALUES (?, ?, ?)`,
+				req.IssuerNameID,
+				shardIdx,
+				req.Until.AsTime(),
+			)
+		} else {
+			_, err = txWithCtx.Exec(
+				`UPDATE crlShards
+				SET leasedUntil = ?
+				WHERE issuerID = ?
+				AND idx = ?`,
+				req.Until.AsTime(),
+				req.IssuerNameID,
+				shardIdx,
+			)
+		}
+		if err != nil {
+			return -1, err
+		}
+
+		return shardIdx, err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &sapb.LeaseCRLShardResponse{
+		IssuerNameID: req.IssuerNameID,
+		ShardIdx:     int64(shardIdx.(int)),
+	}, nil
+
+}
+
+// leaseSpecificCRLShard attempts to lease the crl shard for the given issuer
+// and shard index. It returns an error if the specified shard is already
+// leased.
+func (ssa *SQLStorageAuthority) leaseSpecificCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	res, err := ssa.dbMap.WithContext(ctx).Exec(
+		`UPDATE crlShards
+			SET leasedUntil = ?
+			WHERE issuerID = ?
+			AND idx = ?
+			AND (leasedUntil IS NULL OR leasedUntil < ?)`,
+		req.Until.AsTime(),
+		req.IssuerNameID,
+		req.MinShardIdx,
+		ssa.clk.Now(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	rowsAffected, err := res.RowsAffected()
+	if err != nil {
+		return nil, err
+	}
+	if rowsAffected == 0 {
+		return nil, fmt.Errorf("shard %d for issuer %d already leased", req.MinShardIdx, req.IssuerNameID)
+	}
+	if rowsAffected != 1 {
+		return nil, errors.New("update affected unexpected number of rows")
+	}
+
+	return &sapb.LeaseCRLShardResponse{
+		IssuerNameID: req.IssuerNameID,
+		ShardIdx:     req.MinShardIdx,
+	}, nil
+}
+
+// UpdateCRLShard updates the thisUpdate and nextUpdate timestamps of a CRL
+// shard. It rejects the update if it would cause the thisUpdate timestamp to
+// move backwards. It does *not* reject the update if the shard is no longer
+// leased: although this would be unexpected (because the lease timestamp should
+// be the same as the crl-updater's context expiration), it's not inherently a
+// sign of an update that should be skipped. It does reject the update if the
+// identified CRL shard does not exist in the database (it should exist, as
+// rows are created if necessary when leased).
+func (ssa *SQLStorageAuthority) UpdateCRLShard(ctx context.Context, req *sapb.UpdateCRLShardRequest) (*emptypb.Empty, error) {
+	if core.IsAnyNilOrZero(req.IssuerNameID, req.ThisUpdate, req.NextUpdate) {
+		return nil, errIncompleteRequest
+	}
+
+	res, err := ssa.dbMap.WithContext(ctx).Exec(
+		`UPDATE crlShards
```

Comment:
Add an explicit `LIMIT 1` because we don't ever expect there to be more than 1 row per issuerID and shard index. This method doesn't include some logic check if more than 1 row was updated to throw a warning log either.